source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs

```diff
@@ -12,117 +16,408 @@ namespace Halibut.Tests.ServiceModel
 
-            Exception actualException = null;
-            try
+            var next = await sut.DequeueAsync();
+            next.Should().BeNull();
+        }
+        
+        [Test]
+        [TestCaseSource(typeof(AsyncTestCase))]
+        public async Task QueueAndWait_WhenRequestIsDequeued_ButPollingRequestQueueTimeoutIsReached_ShouldWaitTillRequestRespondsAndClearRequest(bool async)
+        {
+            // Arrange
+            const string endpoint = "poll://endpoint001";
+
+            var sut = new PendingRequestQueueBuilder().WithEndpoint(endpoint).Build();
+            var request = new RequestMessageBuilder(endpoint)
+                .WithServiceEndpoint(seb => seb.WithPollingRequestQueueTimeout(TimeSpan.FromMilliseconds(200)))
+                .Build();
+            var expectedResponse = ResponseMessageBuilder.FromRequest(request).Build();
+
+            // Act
+            var queueAndWaitTask = await StartQueueAndWaitAndWaitForRequestToBeQueued(async, sut, request);
+            var dequeued = await sut.DequeueAsync();
+
+            await Task.Delay(1000, CancellationToken.None);
+
+            sut.ApplyResponse(expectedResponse, request.Destination);
+
+            var response = await queueAndWaitTask;
+
+            // Assert
+            dequeued.Should().NotBeNull().And.Be(request);
+            response.Should().Be(expectedResponse);
+
+            var next = await sut.DequeueAsync();
+            next.Should().BeNull();
+        }
+
+        [Test]
+        [TestCaseSource(typeof(AsyncTestCase))]
+        public async Task QueueAndWait_AddingMultipleItemsToQueueInOrder_ShouldDequeueInOrder(bool async)
+        {
+            // Arrange
+            const string endpoint = "poll://endpoint001";
+
+            var sut = new PendingRequestQueueBuilder().WithEndpoint(endpoint).Build();
+
+            var requestsInOrder = Enumerable.Range(0, 3)
+                .Select(_ => new RequestMessageBuilder(endpoint).Build())
+                .ToList();
+
+            // Act
+            var queueAndWaitTasks = new List<Task<ResponseMessage>>();
+            foreach (var request in requestsInOrder)
             {
-                await task;
+                var queueAndWaitTask = await StartQueueAndWaitAndWaitForRequestToBeQueued(async, sut, request);
+                queueAndWaitTasks.Add(queueAndWaitTask);
             }
-            catch (Exception ex)
+
+            // Assert
+            foreach (var expectedRequest in requestsInOrder)
             {
-                actualException = ex;
+                var request = await sut.DequeueAsync();
+                request.Should().Be(expectedRequest);
             }
 
+            await ApplyResponsesToCompleteWaitingTasks(sut, requestsInOrder, queueAndWaitTasks);
+        }
+
+        [Test]
+        [TestCaseSource(typeof(AsyncTestCase))]
+        public async Task QueueAndWait_AddingMultipleItemsToQueueConcurrently_AllRequestsShouldBeSuccessfullyQueued(bool async)
```

Comment:
Would be worth testing concurrently applying responses.

Does this test that the responses go to the correct requests?

Reply:
> Would be worth testing concurrently applying responses.

Yeah, it would be done!

> Does this test that the responses go to the correct requests?

Actually, I don't think it does. Done!


I was a bit cheeky, and covered both of these requirements in `ApplyResponsesConcurrentlyAndEnsureAllQueueResponsesMatch`. Feel free to change it/split it out into separate test cases if you'd like 🙂